pennylane/beta/interfaces/autograd.py

```diff
@@ -0,0 +1,153 @@
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the mixin interface class for creating differentiable quantum tapes with
+Autograd.
+"""
+# pylint: disable=protected-access
+import autograd.extend
+import autograd.builtins
+
+from pennylane import numpy as np
+
+from pennylane.beta.queuing import AnnotatedQueue
+
+
+class AutogradInterface(AnnotatedQueue):
+    """Mixin class for applying an autograd interface to a :class:`~.QuantumTape`.
+
+    Autograd-compatible quantum tape classes can be created via subclassing:
+
+    .. code-block:: python
+
+        class MyAutogradQuantumTape(AutogradInterface, QuantumTape):
+
+    Alternatively, the autograd interface can be dynamically applied to existing
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
+    differentiable programming.
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitAutograd`, the Autograd interface **does not need to be applied**. It
+        is only applied to tapes executed on non-Autograd compatible devices.
+
+    **Example**
+
+    One an autograd quantum tape has been created, it can be differentiated using autograd:
+
+    .. code-block:: python
+
+        def cost_fn(x, y, z, device):
+            tape = AutogradInterface.apply(QuantumTape())
+
+            with tape:
+                qml.Rot(x, y ** 2, y * np.sin(z), wires=0)
+                expval(qml.PauliX(0))
+
+            return tape.execute(device=device)
+
+    >>> x = np.array(0.1, requires_grad=False)
+    >>> y = np.array(0.2, requires_grad=True)
+    >>> z = np.array(0.3, requires_grad=True)
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> cost_fn(x, y, z, device=dev)
+    [0.03991951]
+    >>> jac_fn = qml.jacobian(cost_fn)
+    >>> jac_fn(x, y, z, device=dev))
+    [[ 0.39828408 -0.00045133]]
+    """
+
+    # pylint: disable=attribute-defined-outside-init
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "autograd"
+
+    def _update_trainable_params(self):
+        params = [o.data for o in self.operations + self.observables]
+        params = [item for sublist in params for item in sublist]
+
+        trainable_params = set()
+
+        for idx, p in enumerate(params):
+            if getattr(p, "requires_grad", True):
+                trainable_params.add(idx)
+
+        self.trainable_params = trainable_params
+        return params

```

Comment:
Should we be returning something and updating the trainable params? It feels like we should just repeat the two lines
```python
        params = [o.data for o in self.operations + self.observables]
        params = [item for sublist in params for item in sublist]
```
within `get_parameters()`.

Reply:
Yes, probably 😆 This was probably me not wanting to repeat those two lines, I have a fear of introducing bottlenecks as we build out the new core

Reply:
@trbromley interesting that `get_parameters()` is the next bottleneck after `_grad_method`. Are these two list comprehensions the culprit, or is there something else I am missing?